pkg/execution/state/redis_state/redis_state.go

```diff
@@ -721,14 +725,20 @@ func (m mgr) Delete(ctx context.Context, i state.Identifier) error {
 		m.kf.Errors(ctx, i),
 		m.kf.RunPauses(ctx, i.RunID),
 	}
+
+	performedDeletion := false
 	for _, k := range keys {
 		cmd := m.r.B().Del().Key(k).Build()
-		if err := m.r.Do(callCtx, cmd).Error(); err != nil {
-			return err
+		result := m.r.Do(callCtx, cmd)
+		if count, _ := result.ToInt64(); count > 0 {
+			performedDeletion = true
+		}
+		if err := result.Error(); err != nil {
+			return false, err
```

Comment:
Looks like `performedDeletion` is true if _any_ of the keys are removed. Seeing as we may have multiple executors hitting this path at the same time for the same run, each could return `1` for a subset and then still run finalization.

Could we change `performedDeletion` to be based on a single key? `m.kf.RunMetadata()` is probably a good one.